### ISSUE TITLE
Improve social & replayability mechanics

### ISSUE THREAD

**Issue by toddmedema** [toddmedema/electrify#46]
Would be great to incorporate more modern game design techniques to boost replayability and how much people want to share the game with their friends.

Some ideas:
- Achievements: encourage folks to log in, store with their account. Example achievements: Complete the first tutorial, complete all tutorials, win each of the scenarios, win at each difficulty level
- Improve leaderboard: Let people enter username, show names on the board, celebrate scores at end of each game
- Perhaps with the above, there's a new main menu item for achievements, that shows your scores, earned badges, and progress towards unearned badges? (like Steam achievements page)
- On the scenarios list page, if you're logged in, indicate which ones you have/haven't played yet (perhaps by showing your best score vs the global high score?)

**Comment by toddmedema** [comment]
## Implementation plan

I traced how the existing pieces (auth, `scores` in Firestore, `plays` in localStorage, scenario list, end-of-game dialog) fit together. Line references are against `master` as of writing.

### Where things stand today

| Piece | Today |
|---|---|
| Auth | Google popup only (`src/Globals.tsx:22`). No sign-out, no display name anywhere. `UserType` is `{ uid?: string }` (`src/Types.tsx:288`) |
| Scores | `addDoc` into Firestore `scores` from inside a reducer (`src/reducers/User.tsx:16`); fields `{score, scoreBreakdown, scenarioId, difficulty, date, uid}` |
| Leaderboard | Two `getDocs` queries in `src/components/views/NewGameDetails.tsx:83`, score + difficulty columns only, **entirely hidden until you log in** |
| Progress | `plays` in localStorage only (`src/LocalStorage.tsx:65`) — device-local, lost on a new browser. Drives tutorial checkmarks in `src/components/views/NewGame.tsx:130` |
| End of game | Dialog with score breakdown (`src/reducers/Game.tsx:481`) — no rank, no PB comparison, no share |
| Firestore rules/indexes | Not in the repo — console-managed |

So this issue is really four features on top of one missing foundation: **a per-account player profile**.

---

### Phase 0 — Foundation: profile document + rules in the repo

**Data model** (one doc per player keeps reads at 1 per login):

```
users/{uid}          { displayName, displayNameLower, achievements: { <id>: <epochMs> },
                       bests: { <scenarioId>: { score, difficulty, date } },
                       createdAt, updatedAt }
usernames/{lower}    { uid }        // create-only claim doc, enforces uniqueness in rules
scores/{auto}        + displayName  // denormalized so the leaderboard is one query, not N+1
```

**Add `firestore.rules`, `firestore.indexes.json`, `firebase.json` to the repo.** Right now rule changes are invisible to reviewers and easy to forget before a deploy — and every phase below needs a rule change. Rules: `users` readable by anyone (it holds nothing private), writable only by owner with `displayName` length/charset validation and achievements append-only; `usernames` create-if-absent only; `scores` create-only where `uid == request.auth.uid`.

Indexes needed: `scores(scenarioId, score desc)` and `scores(scenarioId, uid, score desc)` already exist implicitly — declare them so a fresh project deploys clean.

> Accepted limitation, worth stating in the PR: scores and achievements stay client-authoritative, exactly as today. Rules can't validate a simulation. If cheating on the board becomes real, the fix is a Cloud Function that recomputes the score from a seed + action log — out of scope here.

**Client changes:**

- Extend `UserType` to `{ uid?, displayName?, achievements?, bests?, profileLoaded? }`.
- Convert the `submitHighscore` side-effect-in-a-reducer into a proper thunk in `src/reducers/User.tsx` (`createAsyncThunk`). It needs to `await` its write anyway for the rank lookup in Phase 4, and firing network calls from a reducer is what makes Phase 4 awkward otherwise.
- In `App.tsx`'s `onAuthStateChanged`, dispatch a `loadProfile` thunk: read `users/{uid}`, create it if missing, and **merge local progress up** — localStorage `plays` and locally-earned achievements union into the cloud doc on first login, so nobody loses the tutorials they did logged-out.

### Phase 1 — Display names, so the board has people on it

- First login opens a "Choose your leaderboard name" dialog (seeded from the Google `displayName`, which we currently throw away). 3–20 chars, `[A-Za-z0-9 _-]`, uniqueness claimed via a transaction on `usernames/{lower}`, plus a small blocklist.
- Settings card gets an Account section: signed in as X, change name, **log out** — there's no sign-out in the app at all today, which stops mattering only once accounts carry data.
- On rename, best-effort batch-update the player's own `scores` docs so old rows show the new name.

### Phase 2 — Achievement engine

New `src/data/Achievements.tsx`, declarative and data-driven like `Scenarios.tsx`:

```ts
{ id: "tutorials_all",        // stable string, never renumber -- same rule as scenario ids
  name, description, icon,
  earned: (ctx) => boolean,
  progress?: (ctx) => { current, target } }   // powers the "3 of 6" bars
```

`ctx` = `{ playedScenarioIds, bests, difficultiesWon, lastGameSummary }`.

Starting set, straight from this issue: `tutorial_first`, `tutorials_all`, one `win_<scenarioId>` generated per non-tutorial scenario (so new scenarios add badges for free), and one `difficulty_<Intern…CEO>` per level. Plus 2–3 gameplay badges from data `summarizeHistory` already produces (a blackout-free run, a zero-emission year) to prove the pattern extends.

Evaluation hooks — two places, both already exist:

- tutorial walkthrough completion, `src/components/CompositorContainer.tsx:68`
- scenario survival, `src/reducers/Game.tsx:444`

Both call one `unlockAchievements(ids)` thunk that diffs against what is already earned, writes to localStorage always and to `users/{uid}` (`setDoc` + `merge`) when logged in, and surfaces newly-earned ones via the existing `snackbarOpen` ("🏆 Achievement unlocked: …", collapsed into one message when several land at once).

### Phase 3 — Achievements card

New `ACHIEVEMENTS` in `CardNameType`, `Achievements.tsx` + `AchievementsContainer.tsx`, a `case` in `Compositor.tsx`, and a Main Menu button.

Layout reuses what the tutorial list already does well: "12 of 27 earned" + `LinearProgress`, then `build-list-item` cards with a greyed avatar when locked, description, earned date or a progress bar. Logged-out shows local achievements with a "Log in to save these" CTA — the exact nudge this issue is asking for.

### Phase 4 — Leaderboard + end-of-game celebration

**Board** (`NewGameDetails.tsx`): rank / name / score / difficulty columns, own rows highlighted, "Anonymous" fallback for pre-migration scores. **Make it publicly readable** — hiding the board from logged-out players inverts the incentive; seeing named strangers above you is the reason to log in.

**Celebration** — the actual share lever. After the score write resolves, the victory dialog gains:

- "New personal best — was 640" (from `bests[scenarioId]`)
- "#4 on the global leaderboard" via `getCountFromServer` on `score > mine` — one aggregate read
- any achievements just unlocked
- a **Share** button: Web Share API with a clipboard fallback, text like *"I scored 812 running &lt;scenario&gt; at CEO difficulty on Electrify — electrifygame.com"*

### Phase 5 — Scenario list indicators

`ScenarioListItem` in `NewGame.tsx` shows played/unplayed and "Your best: 812" from `bests`.

**Scope trim, flagged explicitly:** the issue asks for *your best vs the global high score* on the list. Global bests would mean one query per scenario on every list render (12 today, growing). Doing that properly needs a `scenarioStats/{id}` aggregate doc maintained by a Cloud Function — a real backend addition. I would ship your-best-only on the list, keep the global comparison on the details page where it is already loaded, and file the aggregate doc as a follow-up.

---

### Tests

Per CONTRIBUTING, `npm run check` is the gate:

- `src/data/Achievements.test.tsx` — each criterion against synthetic contexts, including the generated per-scenario and per-difficulty ones.
- `src/reducers/User.test.tsx` — unlock dedupes, local/cloud merge on login keeps the union, rename claim collision fails cleanly. Needs a `firebase/firestore` mock; `setupTests.ts` already does this for `firebase/analytics`, so add it alongside.
- `src/LocalStorage.test.tsx` — achievement read/write and the existing `plays` helpers.

### PR sequencing

Five PRs, each shippable: **0** profile doc + rules/indexes in repo + `submitHighscore` → thunk · **1** display names + logout · **2** engine + unlock hooks + snackbar · **3** achievements card · **4** leaderboard names, rank, share · **5** scenario list indicators. Phase 0's rules must be deployed *before* its client code ships, or the profile reads fail closed.

**Main risks:** rules/index drift between repo and console (Phase 0 fixes this permanently); client-authoritative scores (accepted, documented); denormalized names going stale on rename (mitigated, best-effort); and the local-to-cloud merge, which is the one piece where a bug silently eats a player's tutorial progress — hence the tests above.
